File: proyectoDalgoP1/ProblemaP1.java
//Proyecto Realizado por:
//Angela Jimenez - 202210989
//Andrés Felipe Alfonso Gamba - 202210412

import java.io.*;
import java.util.*;
import java.util.stream.IntStream;

public class ProblemaP1 {
    
    public static int maxCarryReal(int k) {
        return (9 * k) / 10;
    }
    
    public static int[] crearTablaScore(int k, int maxT) {
        int[] tabla = new int[maxT + 1];

        for (int T = 0; T <= maxT; T++) {
            int q = T / 3;
            int r = T % 3;
            int s = (q + 2) / 3;
            
            if (r == 0 || s < k) {
                tabla[T] = q;
            } else {
                int m = q % 3;
                tabla[T] = q - (m == 1 ? 1 : 2);
            }
        }
        return tabla;
    }
    
    public static long resolverFestivalRobots(int n, int k, int[] valoresP) {
        int maxT = 9 * k;
        int[] tablaScore = crearTablaScore(k, maxT);
        
        int maxCarry = maxCarryReal(k) + 5;
        
        int[] digitosN;
        if (n == 0) {
            digitosN = new int[]{0};
        } else {
            List<Integer> digitosList = new ArrayList<>();
            int tempN = n;
            while (tempN > 0) {
                digitosList.add(tempN % 10);
                tempN /= 10;
            }
            digitosN = digitosList.stream().mapToInt(i -> i).toArray();
        }
        
        int numDigitos = digitosN.length;
        
        // Extender valores_p con ceros si es necesario
        int[] valoresPExtendidos = new int[numDigitos];
        for (int i = 0; i < numDigitos; i++) {
            valoresPExtendidos[i] = (i < valoresP.length) ? valoresP[i] : 0;
        }
        
        // DP usando arrays: -1 indica estado inalcanzable
        long[] dpPrev = new long[maxCarry + 1];
        Arrays.fill(dpPrev, -1);
        dpPrev[0] = 0;  // Caso base: columna 0, acarreo 0
        
        // Procesar cada columna
        for (int col = 0; col < numDigitos; col++) {
            long[] dpCurr = new long[maxCarry + 1];
            Arrays.fill(dpCurr, -1);
            
            int dP = digitosN[col];
            int PP = valoresPExtendidos[col];
            
            for (int cin = 0; cin <= maxCarry; cin++) {
                if (dpPrev[cin] == -1) {  // Estado inalcanzable
                    continue;
                }
                
                int lo = (cin - dP + 9) / 10;
                if (lo < 0) {
                    lo = 0;
                }
                int hi = Math.min(maxCarry, (9 * k + cin - dP) / 10);
                
                for (int cout = lo; cout <= hi; cout++) {
                    int T = dP - cin + 10 * cout;
                    
                    long ganancia = (long) PP * tablaScore[T];
                    
                    long nuevaCreatividad = dpPrev[cin] + ganancia;
                    if (dpCurr[cout] == -1 || dpCurr[cout] < nuevaCreatividad) {
                        dpCurr[cout] = nuevaCreatividad;
                    }
                }
            }
            
            dpPrev = dpCurr;
        }
        
        return dpPrev[0] != -1 ? dpPrev[0] : 0;
    }
    
    public static void main(String[] args) throws IOException {
        // long tiempoInicio = System.nanoTime();
        
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter pw = new PrintWriter(System.out);
        
        int numCasos = Integer.parseInt(br.readLine().trim());
        
        for (int i = 0; i < numCasos; i++) {
            String[] datos = br.readLine().trim().split(" ");
            int k = Integer.parseInt(datos[0]);  // número de celdas
            int n = Integer.parseInt(datos[1]);  // energía total
            
            // valores P0 P1 P2 P3 P4
            int[] valoresP = new int[datos.length - 2];
            for (int j = 2; j < datos.length; j++) {
                valoresP[j - 2] = Integer.parseInt(datos[j]);
            }
            
            long resultado = resolverFestivalRobots(n, k, valoresP);
            pw.println(resultado);
        }
        
        pw.flush();
        br.close();
        
        // long tiempoTotal = System.nanoTime() - tiempoInicio;
        // System.err.printf("Tiempo: %.6fs%n", tiempoTotal / 1_000_000_000.0);
    }
}

File: proyectoDalgoP1/ProblemaP1.py
#Proyecto Realizado por:
#Angela Jimenez Gonzalez - 202210989
#Andrés Felipe Alfonso Gamba - 202210412

import time

def max_carry_real(k):
    """Calcula el acarreo máximo real posible"""
    return (9 * k) // 10

def crear_tabla_score(k, max_T):
    """Precalcula score_col para todos los T posibles"""
    tabla = [0] * (max_T + 1)  # Usar array para O(1) acceso
    for T in range(max_T + 1):
        q, r = divmod(T, 3)
        s = (q + 2) // 3
        if r == 0 or s < k:
            tabla[T] = q
        else:
            m = q % 3
            tabla[T] = q - (1 if m == 1 else 2)
    return tabla


def resolver_festival_robots(n, k, valores_p):
    # Precalcular tabla de score_col para O(1) lookup
    max_T = 9 * k
    tabla_score = crear_tabla_score(k, max_T)
    
    # Calcular límite preciso de acarreo con margen de seguridad
    max_carry = max_carry_real(k) + 5
    
    # Convertir n a dígitos (de derecha a izquierda: unidades, decenas, ...)
    digitos_n = []
    temp_n = n
    while temp_n > 0:
        digitos_n.append(temp_n % 10)
        temp_n //= 10
    
    # Si n = 0, caso especial
    if not digitos_n:
        digitos_n = [0]
    
    num_digitos = len(digitos_n)
    
    # Extender valores_p con ceros si es necesario
    if len(valores_p) < num_digitos:
        valores_p = valores_p + [0] * (num_digitos - len(valores_p))
    
    # DP usando arrays: -1 indica estado inalcanzable
    dp_prev = [-1] * (max_carry + 1)
    dp_prev[0] = 0  # Caso base: columna 0, acarreo 0
    
    # Procesar cada columna
    for col in range(num_digitos):
        dp_curr = [-1] * (max_carry + 1)
        d_p = digitos_n[col]
        P_p = valores_p[col]
        
        # Para cada acarreo de entrada alcanzable
        for cin in range(max_carry + 1):
            if dp_prev[cin] == -1:  # Estado inalcanzable
                continue
            
            # NUEVA IMPLEMENTACIÓN: Iterar sobre cout en lugar de T
            # Calcular rango de cout válidos usando techo entero
            lo = (cin - d_p + 9) // 10
            if lo < 0:
                lo = 0
            hi = min(max_carry, (9 * k + cin - d_p) // 10)
            
            for cout in range(lo, hi + 1):
                # Calcular T correspondiente (garantizado en rango [0, 9*k])
                T = d_p - cin + 10 * cout
                
                # Calcular ganancia usando tabla precalculada
                ganancia = P_p * tabla_score[T]
                
                # Actualizar DP
                nueva_creatividad = dp_prev[cin] + ganancia
                if dp_curr[cout] == -1 or dp_curr[cout] < nueva_creatividad:
                    dp_curr[cout] = nueva_creatividad
        
        dp_prev = dp_curr
    
    # La respuesta está en dp_curr[0] (acarreo final debe ser 0)
    return dp_prev[0] if dp_prev[0] != -1 else 0


if __name__ == "__main__":
    import sys
    
    tiempo_inicio = time.time()
    
    input = sys.stdin.read
    data = input().strip().split('\n')
    
    num_casos = int(data[0].strip())
    index = 1
    
    for i in range(num_casos):
        # Leer k n P0 P1 P2 P3 P4 (todo en una línea)
        datos = list(map(int, data[index].strip().split()))
        k = datos[0]  # número de celdas
        n = datos[1]  # energía total
        valores_p = datos[2:]  # valores P0 P1 P2 P3 P4
        index += 1
        
        resultado = resolver_festival_robots(n, k, valores_p)
        print(resultado)
    
    tiempo_total = time.time() - tiempo_inicio
    print(f"Tiempo: {tiempo_total:.6f}s", file=sys.stderr)
